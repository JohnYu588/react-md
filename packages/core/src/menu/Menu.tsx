"use client";

import { cnb } from "cnbuilder";
import {
  type CSSProperties,
  type HTMLAttributes,
  forwardRef,
  useEffect,
  useRef,
} from "react";

import { type FloatingActionButtonPosition } from "../button/FloatingActionButton.js";
import { useFocusContainer } from "../focus/useFocusContainer.js";
import { useUserInteractionMode } from "../interaction/UserInteractionModeProvider.js";
import { type ListProps } from "../list/List.js";
import { useAppSize } from "../media-queries/AppSizeProvider.js";
import { type GetDefaultFocusedIndex } from "../movement/types.js";
import { Portal } from "../portal/Portal.js";
import { type CalculateFixedPositionOptions } from "../positioning/types.js";
import {
  type FixedPositioningOptions,
  useFixedPositioning,
} from "../positioning/useFixedPositioning.js";
import { useScrollLock } from "../scroll/useScrollLock.js";
import {
  type ScaleTransitionHookOptions,
  useScaleTransition,
} from "../transition/useScaleTransition.js";
import { type LabelRequiredForA11y, type PropsWithRef } from "../types.js";
import { useEnsuredId } from "../useEnsuredId.js";
import { useIsomorphicLayoutEffect } from "../useIsomorphicLayoutEffect.js";
import {
  type MenuConfiguration,
  MenuConfigurationProvider,
  type MenuOrientationProps,
  useMenuConfiguration,
} from "./MenuConfigurationProvider.js";
import { MenuSheet, type MenuSheetConvenienceProps } from "./MenuSheet.js";
import { MenuWidget } from "./MenuWidget.js";
import { useMenuBarContext } from "./useMenuBarProvider.js";
import { getDefaultAnchor } from "./utils.js";

// NOTE: The augmentation is in this file since no types are imported from the
// `styles` file at this time
declare module "react" {
  interface CSSProperties {
    "--rmd-menu-background-color"?: string;
    "--rmd-menu-color"?: string;
    "--rmd-menu-min-width"?: string | number;
    "--rmd-menu-spacing"?: string | number;
  }
}

const noop = (): void => {
  // do nothing
};

/** @since 5.0.0 */
export type MenuTransitionProps = Omit<
  ScaleTransitionHookOptions<HTMLDivElement>,
  "transitionIn" | "vertical" | "nodeRef"
>;

/**
 * @since 6.0.0
 */
export interface MenuConfigurationProps extends CalculateFixedPositionOptions {
  /**
   * @see {@link ScaleTransitionHookOptions.temporary}
   * @defaultValue `true`
   */
  temporary?: boolean;

  /**
   * @defaultValue `false`
   */
  disablePortal?: boolean;

  /**
   * Boolean if the menu should not gain the elevation styles and should only be
   * set to `true` when rendering within a `Sheet`.
   *
   * @defaultValue `false`
   */
  disableElevation?: boolean;

  /**
   * @defaultValue `false`
   */
  disableTransition?: boolean;

  /**
   * @see {@link FixedPositioningOptions.transformOrigin}
   * @defaultValue `true`
   */
  transformOrigin?: boolean;

  /**
   * Boolean if the menu should close if the page is scrolled. The default
   * behavior is to just update the position of the menu relative to the menu
   * button until it can no longer be visible within the viewport.
   *
   * @defaultValue `false`
   */
  closeOnScroll?: boolean;

  /**
   * Boolean if the page should no longer be scrollable while the menu is
   * visible.
   *
   * @defaultValue `false`
   */
  preventScroll?: boolean;

  /**
   * Boolean if the menu should close instead of repositioning itself if the
   * browser window is resized.
   *
   * @defaultValue `false`
   */
  closeOnResize?: boolean;

  /** @see {@link FixedPositioningOptions.getFixedPositionOptions} */
  getFixedPositionOptions?: () => CalculateFixedPositionOptions;

  /**
   * @defaultValue `false`
   * @see {@link FixedPositioningOptions.disabled}
   */
  disableFixedPositioning?: boolean;
}

/**
 * @since 5.1.0
 * @since 6.0.0 Renamed from `MenuListProps` to `MenuListConvenienceProps`
 */
export interface MenuListConvenienceProps {
  /**
   * An optional style to provide to the `List` component that surrounds the
   * `MenuItem` within a `Menu`.
   */
  listStyle?: CSSProperties;

  /**
   * An optional className to provide to the `List` component that surrounds the
   * `MenuItem` within a `Menu`.
   */
  listClassName?: string;

  /**
   * Any additional props to pass to the `List` component that surrounds the
   * `Menu`'s `MenuItem`s.
   */
  listProps?: PropsWithRef<Omit<ListProps, "horizontal">>;
}

/**
 * @since 6.0.0
 */
export interface MenuConvenienceProps extends MenuConfigurationProps {
  /**
   * This can be used to apply additional props to the `Menu` component.
   *
   * Note: You can override the `style` and `className` using
   * {@link menuStyle} and {@link menuClassName} instead for convenience.
   *
   * @example
   * ```tsx
   * <DropdownMenu
   *   {...props}
   *   menuProps={{
   *     style: {
   *       // custom inline style
   *     },
   *     className: "come-class-name",
   *     getFixedPositionOptions: () => ({
   *       preventOverlap: true,
   *     }),
   *   }}
   * />
   * ```
   */
  menuProps?: PropsWithRef<
    Omit<
      MenuProps,
      | "children"
      | "fixedTo"
      | "visible"
      | "onRequestClose"
      | "getDefaultFocusedIndex"
    >
  >;

  /**
   * Convenience prop to apply custom style to the `Menu` component.
   */
  menuStyle?: CSSProperties;

  /**
   * Convenience prop to apply custom class name to the `Menu` component.
   */
  menuClassName?: string;
}

/**
 * @since 6.0.0
 */
export type MenuFixedPositioningOptions = Omit<
  FixedPositioningOptions<HTMLElement, HTMLDivElement>,
  "onScroll" | "onResize" | "nodeRef" | "disabled"
>;

/**
 * @since 5.0.0
 * @since 6.0.0 Updated to use the latest Menu, Transition, and Portal API.
 */
export interface MenuProps
  extends HTMLAttributes<HTMLDivElement>,
    MenuConfiguration,
    MenuConfigurationProps,
    MenuFixedPositioningOptions,
    MenuOrientationProps,
    MenuTransitionProps,
    MenuListConvenienceProps,
    MenuSheetConvenienceProps {
  visible: boolean;
  onRequestClose: () => void;

  /**
   * @defaultValue `"menu-" + useId()`
   */
  id?: string;

  /**
   * This is used to set the default focus index when the menu is visible.
   *
   * @internal
   */
  getDefaultFocusedIndex?: GetDefaultFocusedIndex;

  /**
   * Custom style that should be applied to the menu only while not rendered
   * within a sheet since the {@link style} would be applied to both versions.
   */
  menuStyle?: CSSProperties;

  /**
   * Custom class name that should be applied only while not rendered within a
   * sheet.
   */
  menuClassName?: string;

  /**
   * @internal
   *
   * This is only used to update the default anchor when the DropdownMenu's
   * toggle is a floating action button.
   */
  floating?: FloatingActionButtonPosition;
}

/**
 * **Client Component**
 *
 * This component should generally only be used to implement context menus with
 * the `useContextMenu` hook. Otherwise, the `DropdownMenu` component should be
 * used.
 *
 * @see The `useContextMenu` hook for an example.
 *
 * @see {@link https://react-md.dev/components/menu | Menu Demos}
 * @since 5.0.0
 * @since 6.0.0 Updated this component to implement all the `Menu`
 * functionality instead of requiring the `useMenu` hook and `MenuWidget`
 * component. In addition, the `renderAsSheet` behavior has been moved into this
 * implementation so that the `MenuRenderer` is no longer required and context
 * menus can appear as a `Sheet`.
 */
export const Menu = forwardRef<HTMLDivElement, LabelRequiredForA11y<MenuProps>>(
  function Menu(props, propRef) {
    const {
      id: propId,
      style: propStyle,
      role = "menu",
      children,
      horizontal: _horizontal,
      sheetHeader: _sheetHeader,
      sheetFooter: _sheetFooter,
      renderAsSheet: _renderAsSheet,
      sheetPosition: _sheetPosition,
      sheetVerticalSize: _sheetVerticalSize,
      sheetProps,
      sheetStyle,
      sheetClassName,
      menuStyle,
      menuClassName,
      disableElevation = false,
      temporary = true,
      tabIndex = -1,
      fixedTo,
      className,
      classNames,
      timeout,
      appear,
      enter,
      exit,
      onEnter,
      onEntering = noop,
      onEntered = noop,
      onExit,
      onExiting,
      onExited = noop,
      onKeyDown = noop,
      listProps,
      listStyle,
      listClassName,
      visible,
      onRequestClose,
      floating,
      anchor,
      closeOnResize = false,
      closeOnScroll = false,
      preventScroll = false,
      vwMargin,
      vhMargin,
      xMargin,
      yMargin,
      width,
      transformOrigin = true,
      preventOverlap,
      disableSwapping,
      disableVHBounds,
      initialX,
      initialY,
      disableFixedPositioning,
      getFixedPositionOptions,
      disablePortal: propDisablePortal,
      disableTransition,
      ...remaining
    } = props;
    const { "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy } =
      props;

    const id = useEnsuredId(propId, "menu");
    const {
      root,
      menubar,
      menuitem,
      activeId,
      animatedOnceRef,
      hoverTimeoutRef,
      disableHoverMode,
    } = useMenuBarContext();
    const {
      horizontal,
      sheetHeader,
      sheetFooter,
      renderAsSheet,
      sheetPosition,
      sheetVerticalSize,
    } = useMenuConfiguration(props);
    const { isPhone } = useAppSize();
    const isSheet =
      renderAsSheet === true || (renderAsSheet === "phone" && isPhone);

    const entered = useRef(false);
    const cancelUnmountFocus = useRef(false);
    const hideWithoutRefocus = (): void => {
      cancelUnmountFocus.current = true;
      onRequestClose();
    };
    const mode = useUserInteractionMode();
    const mouse = mode === "mouse";

    const { eventHandlers, transitionOptions } = useFocusContainer({
      nodeRef: propRef,
      activate: visible,
      onKeyDown(event) {
        onKeyDown(event);

        // when a menu is within a sheet, it should not trigger the custom
        // keyboard behavior
        if (isSheet) {
          return;
        }

        switch (event.key) {
          case "Escape":
            // prevent parent components that have an "Escape" keypress event
            // from being triggered as well
            event.stopPropagation();
            disableHoverMode();
            onRequestClose();
            break;
          case "Tab":
            // since menus are portalled, tab index is kinda broke so just close
            // the menu instead of doing default tab behavior
            event.preventDefault();

            if (!menuitem) {
              // pressing the tab key should still cascade close all menus
              event.stopPropagation();
            }
            disableHoverMode();
            onRequestClose();
            break;
          case "ArrowUp":
            if (!root && menuitem && horizontal) {
              event.stopPropagation();
              event.preventDefault();
              onRequestClose();
            }
            break;
          case "ArrowLeft":
            if (!root && menuitem && !horizontal) {
              event.stopPropagation();
              event.preventDefault();
              onRequestClose();
            }
            break;
        }
      },
      onEntering(appearing) {
        onEntering(appearing);
        entered.current = true;
      },
      onEntered(appearing) {
        onEntered(appearing);
        entered.current = true;
        cancelUnmountFocus.current = false;
        animatedOnceRef.current = true;
      },
      onExited() {
        onExited();
        entered.current = false;
      },
      onExiting,
      disableTransition,
      isFocusTypeDisabled(type) {
        if (role === "listbox") {
          return !isSheet;
        }

        if (type === "keyboard") {
          return isSheet;
        }

        const isHoverDisabled = mouse && hoverTimeoutRef.current === 0;
        if (type === "mount") {
          return isHoverDisabled;
        }

        return (
          isHoverDisabled ||
          cancelUnmountFocus.current ||
          (root && !!activeId && id !== activeId)
        );
      },
    });

    const { ref, style, callbacks, updateStyle } = useFixedPositioning({
      ...transitionOptions,
      disabled: disableFixedPositioning,
      onEnter,
      style: isSheet ? propStyle : menuStyle,
      fixedTo,
      anchor: getDefaultAnchor({
        anchor,
        menubar,
        floating,
        menuitem: !root && menuitem,
        horizontal,
      }),
      vwMargin,
      vhMargin,
      xMargin,
      yMargin,
      width,
      transformOrigin,
      preventOverlap,
      disableSwapping,
      disableVHBounds,
      initialX,
      initialY,
      getFixedPositionOptions,
      onResize: closeOnResize ? hideWithoutRefocus : undefined,
      onScroll(_event, data) {
        if (!data.visible || closeOnScroll) {
          hideWithoutRefocus();
        }
      },
    });
    const { rendered, disablePortal, elementProps } = useScaleTransition({
      nodeRef: ref,
      className: cnb(!isSheet && menuClassName, className),
      transitionIn: visible,
      vertical: !horizontal,
      temporary,
      timeout: isSheet || disableTransition ? 0 : timeout,
      classNames,
      appear,
      enter,
      exit,
      onExit,
      onExiting: transitionOptions.onExiting,
      exitedHidden: true,
      ...callbacks,
    });
    useScrollLock(visible && preventScroll);

    // need to make sure that the useEffect does not refire for hiding on click
    // events because of the `window.requestAnimationFrame`. It'll make it so
    // that menu items that update state are unable to close when clicked
    const hide = useRef(onRequestClose);
    useEffect(() => {
      hide.current = onRequestClose;
    });
    useEffect(() => {
      if (!visible) {
        return;
      }

      const callback = (event: globalThis.MouseEvent): void => {
        // this is required for when the transition is disabled
        if (!entered.current) {
          return;
        }

        // if the user clicks outside of the menu to close it, the toggle button
        // should not be focused. instead the nearest focusable element from the
        // click event should be focused when Tab or Shift + tab is pressed
        cancelUnmountFocus.current =
          !(event.target instanceof HTMLElement) ||
          !event.target.closest(`[role="${role}"]`);

        // this won't be called if `event.stopPropagation()` is called
        hide.current();
        disableHoverMode();
      };

      // wait an animation frame so the initial click event that caused the menu
      // to become visible does not immediately close the menu
      const frame = window.requestAnimationFrame(() => {
        window.addEventListener("click", callback);
      });

      return () => {
        window.cancelAnimationFrame(frame);
        window.removeEventListener("click", callback);
      };
    }, [disableHoverMode, role, visible]);
    useIsomorphicLayoutEffect(() => {
      if (!visible) {
        return;
      }

      updateStyle();
    }, [updateStyle, children, visible]);

    return (
      <MenuConfigurationProvider
        horizontal={horizontal}
        renderAsSheet={renderAsSheet}
        sheetFooter={sheetFooter}
        sheetHeader={sheetHeader}
        sheetPosition={sheetPosition}
        sheetVerticalSize={sheetVerticalSize}
      >
        <MenuSheet
          aria-label={ariaLabel}
          aria-labelledby={ariaLabelledBy as string}
          header={sheetHeader}
          footer={sheetFooter}
          position={sheetPosition}
          verticalSize={sheetVerticalSize}
          visible={visible}
          enabled={isSheet}
          onRequestClose={onRequestClose}
          style={sheetStyle}
          className={sheetClassName}
          disablePortal={propDisablePortal}
          temporary={temporary}
          disableTransition={disableTransition}
          {...sheetProps}
        >
          <Portal disabled={isSheet || (propDisablePortal ?? disablePortal)}>
            {(rendered || isSheet) && (
              <MenuWidget
                {...remaining}
                {...elementProps}
                {...eventHandlers}
                id={id}
                role={role}
                style={isSheet ? propStyle : style}
                isSheet={isSheet}
                tabIndex={tabIndex}
                horizontal={horizontal}
                listProps={listProps}
                listStyle={listStyle}
                listClassName={listClassName}
                disableElevation={disableElevation}
                cancelUnmountFocus={cancelUnmountFocus}
              >
                {children}
              </MenuWidget>
            )}
          </Portal>
        </MenuSheet>
      </MenuConfigurationProvider>
    );
  }
);
