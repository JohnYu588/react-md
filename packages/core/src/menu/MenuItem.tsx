"use client";
import { cnb } from "cnbuilder";
import { forwardRef } from "react";
import type { ListItemProps } from "../list/ListItem.js";
import { ListItem } from "../list/ListItem.js";
import { useKeyboardMovementContext } from "../movement/useKeyboardMovementProvider.js";
import { useEnsuredId } from "../useEnsuredId.js";

/**
 * @remarks \@since 5.0.0
 */
export interface MenuItemProps extends Omit<ListItemProps, "role"> {
  /**
   * @defaultValue `"menuitem"`
   */
  role?: string;

  /**
   * @defaultValue `-1`
   */
  tabIndex?: number;
}

/**
 * **Client Component**
 *
 * This component is used as an "action" within a `Menu`/`DropdownMenu` that
 * implements some keyboard focus behavior. This component should generally have
 * an `onClick` event handler.
 *
 * @remarks \@since 5.0.0
 */
export const MenuItem = forwardRef<HTMLLIElement, MenuItemProps>(
  function MenuItem(props, ref) {
    const {
      id: propId,
      role = "menuitem",
      tabIndex: propTabIndex,
      children,
      className,
      ...remaining
    } = props;

    const id = useEnsuredId(propId, "menuitem");
    const { tabIndexBehavior, activeDescendantId } =
      useKeyboardMovementContext();
    const focused = id === activeDescendantId;
    let tabIndex = propTabIndex ?? -1;
    if (tabIndexBehavior === "roving" && focused) {
      tabIndex = 0;
    }

    return (
      <ListItem
        {...remaining}
        id={id}
        ref={ref}
        role={role}
        tabIndex={tabIndex}
        className={cnb(
          "rmd-menu-item",
          tabIndexBehavior === "virtual" && focused && "rmd-menu-item--focused",
          className
        )}
      >
        {children}
      </ListItem>
    );
  }
);
