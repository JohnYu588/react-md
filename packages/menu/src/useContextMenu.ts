import type {
  InitialCoords,
  PositionAnchor,
  UseStateObject,
} from "@react-md/core";
import { BELOW_INNER_LEFT_ANCHOR } from "@react-md/core";
import type { MouseEvent, RefObject } from "react";
import { useCallback, useRef, useState } from "react";

/** @remarks \@since 6.0.0 */
export interface ContextMenuProps extends InitialCoords {
  "aria-label": string;
  anchor: PositionAnchor;
  fixedTo: RefObject<HTMLElement>;
  visible: boolean;
  preventScroll: boolean;
  onRequestClose(): void;
}

/**
 * @remarks
 * \@since 5.0.0
 * \@since 6.0.0  Dropped most options since they are no longer required for the
 * context menu to work. Apply any `Menu` props directly to the `Menu` component
 * instead.
 */
export interface ContextMenuHookOptions {
  /**
   * @defaultValue `BELOW_INNER_LEFT_ANCHOR`
   * @see {@link BELOW_INNER_LEFT_ANCHOR}
   */
  anchor?: PositionAnchor;

  /**
   * @defaultValue `"Context Menu"`
   */
  menuLabel?: string;

  /**
   * @defaultValue `true`
   */
  preventScroll?: boolean;
  onContextMenu?<E extends HTMLElement>(event: MouseEvent<E>): void;
}

/**
 * @remarks
 * \@since 5.0.0
 * \@since 6.0.0 Renamed from `ContextMenuHookReturnValue` to
 * `ContextMenuImplementation` and dropped the `menuRef` and `menuNodeRef`
 * fields.
 */
export interface ContextMenuImplementation
  extends UseStateObject<"visible", boolean> {
  menuProps: ContextMenuProps;
  onContextMenu<E extends HTMLElement>(event: MouseEvent<E>): void;
}

const noop = (): void => {
  // do nothing
};

/**
 * This hook controls the visibility and positioning for a context menu.
 *
 * @example
 * Simple Example
 * ```tsx
 * import type { ReactElement } from "react";
 * import [ Menu, MenuItem, useContextMenu ] from "@react-md/menu":
 *
 * function Example(): ReactElement {
 *   const { menuProps, onContextMenu } = useContextMenu();
 *
 *   return (
 *     <div onContextMenu={onContextMenu}>
 *       <textarea />
 *       <Menu {...menuProps}>
 *         <MenuItem>Cut</MenuItem>
 *         <MenuItem>Copy</MenuItem>
 *         <MenuItem>Paste</MenuItem>
 *         <MenuItem>Undo</MenuItem>
 *       </Menu>
 *     </div>
 *   );
 * }
 * ```
 *
 * @remarks
 * \@since 5.0.0
 * \@since 6.0.0 No longer supports overriding most of the `Menu` props. The
 * props must be passed to the `Menu` component manually.
 */
export function useContextMenu(
  options: ContextMenuHookOptions = {}
): ContextMenuImplementation {
  const {
    anchor = BELOW_INNER_LEFT_ANCHOR,
    menuLabel = "Context Menu",
    onContextMenu = noop,
    preventScroll = true,
  } = options;
  const [coords, setCoords] = useState<InitialCoords>({});
  const [visible, setVisible] = useState(false);
  const fixedTo = useRef<HTMLElement>(null);
  const onRequestClose = useCallback(() => {
    setVisible(false);
  }, []);

  return {
    visible,
    setVisible,
    menuProps: {
      "aria-label": menuLabel,
      anchor,
      ...coords,
      fixedTo,
      visible,
      onRequestClose,
      preventScroll,
    },
    onContextMenu(event) {
      onContextMenu(event);
      if (event.isPropagationStopped()) {
        return;
      }

      event.preventDefault();
      event.stopPropagation();
      setCoords({
        initialX: event.clientX,
        initialY: event.clientY,
      });
      setVisible(true);
    },
  };
}
