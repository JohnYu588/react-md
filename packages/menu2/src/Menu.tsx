import type { ReactElement, RefObject } from "react";
import {
  ConditionalPortal,
  RenderConditionalPortalProps,
} from "@react-md/portal";
import {
  TransitionCallbacks,
  useFixedPositioning,
  useScaleTransition,
} from "@react-md/transition";
import { TOP_INNER_RIGHT_ANCHOR } from "@react-md/utils";

import { KeyboardFocusProvider } from "./KeyboardFocusProvider";
import { MenuWidgetProps, MenuWidget } from "./MenuWidget";

/**
 * @remarks \@since 4.0.0
 */
export interface MenuProps
  extends MenuWidgetProps,
    RenderConditionalPortalProps,
    TransitionCallbacks {
  menuRef: RefObject<HTMLElement>;
  toggleRef: RefObject<HTMLElement>;
  visible: boolean;

  /**
   * Boolean if the `toggleRef` should not be focused once the menu is closed if
   * a MenuItem's action was to focus another element on the page. This should
   * generally be kept at `false` for keyboard accessibility.
   *
   * @defaultValue `false`
   */
  disableToggleFocus?: boolean;
}

export function Menu({
  portal = true,
  portalInto,
  portalIntoId,
  style: propStyle,
  className,
  menuRef,
  toggleRef,
  visible,
  onEnter,
  onEntering,
  onEntered,
  onExit,
  onExiting,
  onExited,
  horizontal,
  disableToggleFocus = false,
  ...props
}: MenuProps): ReactElement {
  const { style, transitionOptions } = useFixedPositioning({
    style: propStyle,
    nodeRef: menuRef,
    fixedTo: toggleRef,
    anchor: TOP_INNER_RIGHT_ANCHOR,
    onEnter,
    onEntered,
    onEntering(appearing) {
      onEntering?.(appearing);
      menuRef.current?.focus();
    },
    onExited,
    transformOrigin: true,
  });
  const { elementProps, rendered } = useScaleTransition({
    ...transitionOptions,
    className,
    transitionIn: visible,
    vertical: !horizontal,
    onExit,
    onExiting() {
      onExiting?.();
      if (!disableToggleFocus) {
        toggleRef.current?.focus();
      }
    },
  });
  return (
    <ConditionalPortal
      portal={portal}
      portalInto={portalInto}
      portalIntoId={portalIntoId}
    >
      {rendered && (
        <KeyboardFocusProvider>
          <MenuWidget {...props} {...elementProps} style={style} />
        </KeyboardFocusProvider>
      )}
    </ConditionalPortal>
  );
}
