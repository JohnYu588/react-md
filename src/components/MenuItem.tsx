/* eslint-disable */
import { RippleContainer, useElementInteraction } from "@react-md/core";
import { cnb } from "cnbuilder";
import type { HTMLAttributes, ReactElement } from "react";

import styles from "./MenuItem.module.scss";

export type MenuItemProps = HTMLAttributes<HTMLDivElement>;

export function MenuItem(props: MenuItemProps): ReactElement {
  const { "aria-disabled": disabled, className, ...remaining } = props;
  const { pressed, handlers, rippleContainerProps } = useElementInteraction({
    onClick(event) {
      console.log("CLICK");
    },
    disabled: disabled === "false" || disabled === true,

    ...remaining,
  });

  return (
    <div
      {...remaining}
      {...handlers}
      className={cnb(styles.container, className)}
      tabIndex={0}
      role="menuitem"
    >
      MenuItem
      {rippleContainerProps && <RippleContainer {...rippleContainerProps} />}
    </div>
  );
}
