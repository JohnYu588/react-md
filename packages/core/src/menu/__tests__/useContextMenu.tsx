import { describe, expect, it } from "@jest/globals";
import { type ReactElement } from "react";

import { fireEvent, rmdRender, screen } from "../../test-utils/index.js";
import { Menu } from "../Menu.js";
import { MenuItem } from "../MenuItem.js";
import { useContextMenu } from "../useContextMenu.js";

function Test(): ReactElement {
  const { menuProps, onContextMenu } = useContextMenu();
  return (
    <>
      <label>
        Example
        <textarea onContextMenu={onContextMenu} />
      </label>
      <Menu {...menuProps}>
        <MenuItem>Cut</MenuItem>
        <MenuItem>Copy</MenuItem>
        <MenuItem>Paste</MenuItem>
      </Menu>
    </>
  );
}

describe("useContextMenu", () => {
  // I should add some real tests later...
  it("should work as expected", async () => {
    rmdRender(<Test />);
    const textarea = screen.getByRole("textbox", { name: "Example" });

    fireEvent.contextMenu(textarea, {
      clientX: 300,
      clientY: 600,
    });

    const menu = await screen.findByRole("menu", { name: "Context Menu" });
    // mostly verifying that the useFixedPositioning allows for the initialX and
    // initialY set by clientX and clientY
    expect(menu).toMatchSnapshot();
  });
});
