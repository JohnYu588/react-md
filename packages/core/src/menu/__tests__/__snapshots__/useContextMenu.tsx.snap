// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useContextMenu should work as expected 1`] = `
<div
  aria-label="Context Menu"
  class="rmd-menu rmd-menu--elevated"
  id="menu-:r0:"
  role="menu"
  style="left: 300px; top: 600px; position: fixed; transform-origin: 0 0;"
  tabindex="-1"
>
  <ul
    class="rmd-list"
    role="none"
  >
    <li
      class="rmd-list-item rmd-list-item--normal rmd-interaction-surface rmd-menu-item"
      id="menuitem-:r1:"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="rmd-list-item__text"
      >
        Cut
      </span>
    </li>
    <li
      class="rmd-list-item rmd-list-item--normal rmd-interaction-surface rmd-menu-item"
      id="menuitem-:r2:"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="rmd-list-item__text"
      >
        Copy
      </span>
    </li>
    <li
      class="rmd-list-item rmd-list-item--normal rmd-interaction-surface rmd-menu-item"
      id="menuitem-:r3:"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="rmd-list-item__text"
      >
        Paste
      </span>
    </li>
  </ul>
</div>
`;
